crates/sui-axelar-cgp/move/sources/gateway.move

```diff
@@ -0,0 +1,415 @@
+/// - Module developer on the Sui side will have to implement a system to support messaging;
+/// - Checks for uniqueness of messages should be done through `Channel`s to avoid big data storage;
+///
+/// I. Sending messages
+///
+/// A message is sent through the `send` function, a Channel is supplied to determine the source -> ID.
+/// Event is then emitted and Axelar network can operate
+///
+/// II. Receiving messages
+///
+/// Message bytes and signatures are passed into `create` function to generate a Message object.
+///  - Signatures are checked against the known set of validators.
+///  - Message bytes are parsed to determine: source, destination_chain, payload and target_id
+///  - `target_id` points to a `Channel` object
+///
+/// Once created, `Message` needs to be consumed. And the only way to do it is by calling
+/// `consume_message` function and pass a correct `Channel` instance alongside the `Message`.
+///  - Message is checked for uniqueness (for this channel)
+///  - Message is checked to match the `Channel`.id
+///
+module axelar::gateway {
+    use std::vector;
+
+    use axelar::axelar;
+    use axelar::axelar::{Axelar, validate_proof};
+    use axelar::messaging;
+    use axelar::messaging::Message;
+    use axelar::utils::to_sui_signed;
+    use sui::bcs;
+
+    #[test_only]
+    use axelar::utils::operators_hash;
+    #[test_only]
+    use sui::vec_map;
+
+    /// For when message signatures failed verification.
+    const ESignatureInvalid: u64 = 1;
+
+    /// For when number of commands does not match number of command ids.
+    const EInvalidCommands: u64 = 4;
+
+    /// For when message chainId is not SUI.
+    const EInvalidChain: u64 = 3;
+
+    // These are currently supported
+    const SELECTOR_APPROVE_CONTRACT_CALL: vector<u8> = b"approveContractCall";
+    const SELECTOR_TRANSFER_OPERATORSHIP: vector<u8> = b"transferOperatorship";
+
+    /// The main entrypoint for the external message processing.
+    /// Parses data and attaches messages to the Axelar object to be
+    /// later picked up and consumed by their corresponding Channel.
+    public fun process_messages(
+        axelar: &mut Axelar,
+        input: vector<u8>
+    ) {
+        let messages = validate_and_create_messages(axelar, input);
+        let (i, len) = (0, vector::length(&messages));
+
+        while (i < len) {
+            axelar::add_message(axelar, vector::pop_back(&mut messages));
+            i = i + 1;
+        };
+        vector::destroy_empty(messages);
+    }
+
+    /// Processes the data and the signatures generating a vector of
+    /// `Message` objects.
+    ///
+    /// Aborts with multiple error codes, ignores messages which are not
+    /// supported by the current implementation of the protocol.
+    ///
+    /// Input data must be serialized with BCS (see specification here: https://github.com/diem/bcs).
+    fun validate_and_create_messages(validators: &mut Axelar, input: vector<u8>): vector<Message> {
```

Comment:
messages coming from axelar are `commands`
messages going to into destination channel are `call_approvals`.